### PR TITLE
Updating to use c++20

### PR DESCRIPTION
**Pull request by clemahieu**
Fixing various type ambiguities flagged in c++20 mode.

### CODE REVIEW

File: nano/lib/numbers.cpp [hunk 1]
```diff
@@ -144,6 +144,11 @@ bool nano::public_key::decode_account (std::string const & source_a)
 	return error;
 }
 
+bool nano::public_key::operator== (nano::public_key const &) const
+{
+	return true;
```
Comment (dsiganos) [inline nano/lib/numbers.cpp]: This can't be right.

**Review by dsiganos — CHANGES_REQUESTED** [review]

**Review by dsiganos — DISMISSED** [review]

**Comment by clemahieu** [comment]
Looks like boost will need a new version.

Boost asio coroutine support: https://www.boost.org/users/history/version_1_74_0.html
c++20 compiler testing: https://www.boost.org/users/history/version_1_77_0.html

**Comment by guilhermelawless** [comment]
Stumbled upon here and remembered this one :)

https://github.com/nanocurrency/nano-node/blob/9bb89ab3232eb603911ae1125a78921ec2d4f591/nano/node/distributed_work_factory.cpp#L57

**Comment by thsfs** [comment]
This change requires updating the documentation website, switching the compiler version from C++17 to C++20.

**Pull request merged**